server: emergency file-based authentication for the HTTP APIs + cookie-as-password authentication

knz opened the pull request:
Fixes #29270. 
Fixes #26283. 
Fixes #59376. 

See the individual commits for details.

TLDR 2nd commit:

- it becomes possible to generate an emergency session cookie using the new CLI command `cockroach auth-session emergency-login`. The command also emits a rule to be added to a server-side config file before the cookie can be used. This does not require SQL or KV operations.
- the fact that the rule must be added to a file server-side guarantees that the facility can only be used by administrators.
- using the HTTP interfaces using a cookie created this way will not need SQL or KV to validate the session and whether it has the administrator bit.

TLDR 3rd commit: it's now possible to copy-paste the encoded session cookie as a *password* in the interactive login form in the web browser, for convenience.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60723)
<!-- Reviewable:end -->

knz committed:
server: pre-populate the isAdmin field in HTTP sessions

This commit ensures that the admin bit (whether the logged-in user is
an admin or not) is pre-populated at the moment a HTTP request comes
in. The bit is then stored in the context alongside the username
and session ID, and forwarded during RPC fan-out.

This ensures that:

1. subsequent calls to `hasAdminRole()` throughout the API/RPC code
   reuse the now-cached value. This may save up on SQL/KV latencies.
2. during a fan-out, all the nodes except the gateway can reuse
   the cached value and can avoid a SQL/KV lookup to re-determine
   the admin bit.

Release note: None

knz committed:
cli,server: new emergency authentication feature for HTTP request

Prior to this patch, if a cluster started to malfunction or some
system ranges became unavailable (in particular, any of
`system.users`, `.role_options`, `.role_membership` and
`.web_sessions`), it was not possible to access the HTTP interfaces
any more because authentication would hang or fail.

This was particularly problematic because we want access to the debug
APIs to troubleshoot malfunctioning clusters.

This patch fixes this. See the release note for details.

Release note (security update): It is now possible to re-gain access
to the HTTP APIs, including the debug APIs, when a cluster is
unavailable. To achieve this, a new CLI utility `cockroach
auth-session emergency-login` is introduced. This can be used as
follows:

1. run the utility, this produces an **authentication rule** and a
   separate **login cookie**.
2. copy the **authentication rule** into the emergency login
   file (details below). Send a SIGHUP signal to the process to
   reload the file and activate the login cookie.
3. use the **login cookie** in the HTTP client to access
   the HTTP interface.

The emergency login file is a file read by the CockroachDB upon
startup and when receiving the SIGHUP signal. By default, the file is
named `emergency_sessions` inside the first store (data)
directory. The location of this file can be customized using the (new)
flag `--emergency-authentication-sessions` to the `start` and
`start-single-node` subcommands.

In a Kubernetes / container environment, it is suggested to make
the flag point to a file inside a k8s secret volume mount.

An event of type `emergency_auth_rule` is logged on the OPS channel
whenever a new rule is loaded from the emergency session file.

Reloading the emergency session file (e.g. using a signal) while it is
empty, non-existent or contains invalid/unrecognized data will cause
all existing session tokens to become invalidated.

It is possible to disable this mechanism entirely by passing an empty
string to the new command-line flag.

knz committed:
server: accept session cookies as password during login

Release note (security update): The HTTP log-in form and `/login` API
now recognizes previously-issues session cookies as valid passwords.

This enables the following use cases:

- it makes it possible to conveniently copy-paste the output of
  `cockroach auth-session login` and `cockroach auth-session
  emergency-login` into a web browser. Previously, reusing
  a cookie issued on the command line in a web browser required
  advanced expertise with the browser's development console.

- thanks to the previous benefit, it becomes more convenient for an
  administrator to log in "as" an existing user, without knowing their
  password. They can simply issue `cockroach auth-session login` as
  the desired user and the "session cookie" becomes a temporary
  password to log in as the target user interactively.

knz commented:
@bdarnell can you step in for Aaron who is on PTO and run a security check on this approach? thanks

@tbg can you check whether the facility provided here is a suitable solution to the problem you've enunciated in the two linked issues? thanks

stevendanna reviewed (commented):
I made it through the first two commits and left some comments.  In general I like the idea of an emergency auth mechanism and I think relying on "able to modify files on the server == admin" makes sense to me.

Sometimes for features like this, I think it is useful to remind the user that it is turned on so that they remember to go clean it up after the emergency is over.  I wonder if we could eventually have some indication in the UI that emergency auth had been enabled.

I'll try to give this more time early next week.

> Reviewed 5 of 5 files at r1, 17 of 17 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-:-MWjFUbd8aQhKskCKO60:b-7qvsu3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, and @stevendanna)

___
*[pkg/cli/flags.go, line 1069 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj6R1H-H2jzBbuH8-D:-MWj6R1H-H2jzBbuH8-E:b-6jfd25) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/cli/flags.go#L1069)):*
> ```Go
> 	// Fill in the default for the emergency session file if not passed on the command line.
> 	if !changed(fs, cliflags.EmergencyAuthenticationSessions.Name) {
> 		// Search for the first store with a directory.
> ```

nit: "Search for the first on-disk store" or "Search for the first non-in-memory store" might be more clear. At first, I thought this would be looking for the first store that actually contained the given directory, but that doesn't seem to be the case here. 
___
*[pkg/server/authentication.go, line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj4gDk1YBjrWvzWPsZ:-MWj4gDl333x8n2mJnbr:b-q4xkq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9079604e145bb0c83d4203ce50f30239b627a534/pkg/server/authentication.go#L351)):*
> ```Go
> func (s *authenticationServer) verifySession(
> 	ctx context.Context, cookie *serverpb.SessionCookie,
> ) (isValid bool, user security.SQLUsername, isAdmin bool, err error) {
> ```

It's fine as-is but I wonder if it would make sense to pass back a small struct here?
___
*[pkg/server/file_based_authentication.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjEfhn6h3asGBlG7lJ:-MWjEfhn6h3asGBlG7lK:b99dcp0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L54)):*
> ```Go
> // (only admins can change command-line flag); and if that flag is not
> // present in a file stored in the first store data directory (only
> // admins can write to the store directory).
> ```

I need to look into the "multi-tenant" stuff to see if that would impact any of these assumptions in a meaningful way. I'm assuming not.
___
*[pkg/server/file_based_authentication.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj9XZeACGpsDF6pk6P:-MWj9XZeACGpsDF6pk6Q:bnw4zdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L175)):*
> ```Go
> 			continue
> 		}
> 		parts := strings.Split(line, " ")
> ```

Most users are going to copy/and paste the output and I suppose this is a place to be strict rather than permissive, but if you wanted you could use `strings.Fields(line)` here (handles multiple spaces between entries and other whitespace characters).
___
*[pkg/server/file_based_authentication.go, line 234 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjAMR9C2Mv346KZMl8:-MWjAMR9C2Mv346KZMl9:b-urexud) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L234)):*
> ```Go
> 			IsAdmin:   isAdmin,
> 			ExpiresAt: nanos,
> 		})
> ```

Will this end up creating multiple events for the same auth rule if a user add a session, reloads, adds another session, and reloads again?
___
*[pkg/server/file_based_authentication.go, line 248 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjC_qmEOnzvnRvWwlQ:-MWjC_qmEOnzvnRvWwlR:bp35s3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L248)):*
> ```Go
> 	username := user.Normalized()
> 
> 	expiryTime := timeutil.Now().Add(duration)
> ```

Since the current max duration is hard coded, it might be nice to return an error here if someone is trying to generate a token that we know will fail later on?
___
*[pkg/server/file_based_authentication.go, line 250 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjDgZrAeADbpIOMep_:-MWjDgZrAeADbpIOMepa:bdjx79y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L250)):*
> ```Go
> 	expiryTime := timeutil.Now().Add(duration)
> 
> 	id := int64(rand.Uint64())
> ```

Any reason not to use crypto/rand here? This function should not be called often.
___
*[pkg/server/file_based_authentication.go, line 261 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjEIRC3Tg3GPqSRma0:-MWjEIRC3Tg3GPqSRma1:bustu7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L261)):*
> ```Go
> 	}
> 
> 	authLine := fmt.Sprintf("%s %v %d %s %d",
> ```

nit: the output will be the same, but we can use `%t` for the second argument since we know it is a boolean.
___
*[pkg/server/status.go, line 2412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj55N2-kdBEyRd22uJ:-MWj55N2-kdBEyRd22uK:b-4lohxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9079604e145bb0c83d4203ce50f30239b627a534/pkg/server/status.go#L2412)):*
> ```Go
> 		// If the incoming context doesn't have the metadata fields,
> 		// it's a gRPC / internal SQL connection which has root on the cluster.
> 		return true, true, nil
> ```

This case is probably worth triple checking in our heads that it is safe to return true here.


<!-- Sent from Reviewable.io -->

knz commented:
> so that they remember to go clean it up after the emergency is over

FWIW that's what the expiration timestamp is about.

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60723)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @stevendanna)

___
*[pkg/cli/flags.go, line 1069 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj6R1H-H2jzBbuH8-D:-MWyMkvdBTyB9LyqquNX:blas5yh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/cli/flags.go#L1069)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

nit: "Search for the first on-disk store" or "Search for the first non-in-memory store" might be more clear. At first, I thought this would be looking for the first store that actually contained the given directory, but that doesn't seem to be the case here. 
</blockquote></details>

Agreed. Fixed.
___
*[pkg/server/authentication.go, line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj4gDk1YBjrWvzWPsZ:-MWyMxDR6OemF0IATGul:bcy8g32) ([raw file](https://github.com/cockroachdb/cockroach/blob/9079604e145bb0c83d4203ce50f30239b627a534/pkg/server/authentication.go#L351)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

It's fine as-is but I wonder if it would make sense to pass back a small struct here?
</blockquote></details>

I don't know the answer to this question. We have precedent for using multiple return values in other places in this code already, but I am not opposed to adding grouping structs where it makes sense. The error has to remain separate however, and the content of the struct would be undefined if isValid is false. So we'd use a struct just to group `user` and `isAdmin`?

Let me know what you think is valuable.
___
*[pkg/server/file_based_authentication.go, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjEfhn6h3asGBlG7lJ:-MWyO0xY0H7jiQCzoy_f:b-abg957) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L54)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I need to look into the "multi-tenant" stuff to see if that would impact any of these assumptions in a meaningful way. I'm assuming not.
</blockquote></details>

Yes your assumption is right. Added an explanation to clarify.
___
*[pkg/server/file_based_authentication.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj9XZeACGpsDF6pk6P:-MWyO3pg8pPhI2KLfPTd:b-h990k8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L175)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Most users are going to copy/and paste the output and I suppose this is a place to be strict rather than permissive, but if you wanted you could use `strings.Fields(line)` here (handles multiple spaces between entries and other whitespace characters).
</blockquote></details>

Good call. Fixed.
___
*[pkg/server/file_based_authentication.go, line 234 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjAMR9C2Mv346KZMl8:-MWyODEz9tREOlAPnucD:b-5f024n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L234)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Will this end up creating multiple events for the same auth rule if a user add a session, reloads, adds another session, and reloads again?
</blockquote></details>

Yes it would with the current implementation. Is that an issue? I am not envisioning this function to be called on a periodic basis. Added a comment to explain.
___
*[pkg/server/file_based_authentication.go, line 248 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjC_qmEOnzvnRvWwlQ:-MWySTky-4i0Wo2I96nq:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L248)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Since the current max duration is hard coded, it might be nice to return an error here if someone is trying to generate a token that we know will fail later on?
</blockquote></details>

Good idea. Done.
___
*[pkg/server/file_based_authentication.go, line 250 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjDgZrAeADbpIOMep_:-MWyOwL4EH1L7ZJzciTI:bk0goeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L250)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Any reason not to use crypto/rand here? This function should not be called often.
</blockquote></details>

There is no such reason, but can you explain what the added value would be? The main entropy for this secret is carried by the uuid.NewV4; session Ids are not meant to be very secret, we just expect them to not conflict.

(Which reminds me, we have some risk now here that the session ID generated here will conflict with one in `system.web_sessions`. The UX result of this is that a user who legitimately logs in using a password an a web_sessions cookie can find themselves unable to use that cookie because it doesn't match a per-registered emergency cookie with the same session ID. Do we want to care about this?)
___
*[pkg/server/file_based_authentication.go, line 261 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjEIRC3Tg3GPqSRma0:-MWyPUFg0cD3C-Ogl1M8:bo3vl6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L261)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

nit: the output will be the same, but we can use `%t` for the second argument since we know it is a boolean.
</blockquote></details>

Thanks, done.
___
*[pkg/server/status.go, line 2412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj55N2-kdBEyRd22uJ:-MWyP_Dw8GVPiJ2r6vfs:bf57l54) ([raw file](https://github.com/cockroachdb/cockroach/blob/9079604e145bb0c83d4203ce50f30239b627a534/pkg/server/status.go#L2412)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

This case is probably worth triple checking in our heads that it is safe to return true here.
</blockquote></details>

As the livestream today demonstrated, we have numerous unit tests that fail when we _mistakenly_ return true here (i.e. we have tests that assert that non-admin sessions don't have access to admin-only stuff).

(The return true happens to be OK here,  but only as long as this function is not called from the wrong place...)


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> Reviewed 5 of 5 files at r1, 17 of 17 files at r2, 10 of 10 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60723)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, and @stevendanna)

___
*[pkg/cli/auth.go, line 256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWf-OuP4kQbZ_RsydlB:-MWf-OuP4kQbZ_RsydlC:b-ua2mmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/cli/auth.go#L256)):*
> ```Go
> 	}
> 
> 	expiresAt, authLine, httpCookie, err := server.GenerateEmergencyCredentials(username, true /* isAdmin */, authCtx.validityPeriod)
> ```

Does the `true /* isAdmin */` mean that this token will have admin privileges even if the underlying user does not normally? That seems bad. I understand that in an emergency situation we may not be able to ask the database whether the user is an admin but I think we should require an `--admin` flag on the `emergency-login` flag instead of automatically escalating the given user to an admin. 
___
*[pkg/cli/auth.go, line 264 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWzg_o74scoDT82X8li:-MWzg_o74scoDT82X8lj:b9xm1y9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fd72fa88fb161165002333d98e74c5502ceac19/pkg/cli/auth.go#L264)):*
> ```Go
> 	}
> 
> 	fmt.Println("# entry you must add to the emergency login file on the server:")
> ```

I assume the intention of the leading `#` characters on all non-blank lines is that you could put the entire output of the `emergency-login` command in the file. Would be nice to have an end-to-end (tcl?) test showing that that works and there's no extraneous garbage on stdout. 
___
*[pkg/cli/flags.go, line 1069 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWf0N23DOKKjf0PGVRl:-MWf0N244ruaJegaurvF:b2ems61) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/cli/flags.go#L1069)):*
> ```Go
> 	// Fill in the default for the emergency session file if not passed on the command line.
> 	if !changed(fs, cliflags.EmergencyAuthenticationSessions.Name) {
> 		// Search for the first store with a directory.
> ```

I suggest making this default to the certs directory instead of the first store directory - that directory already has similar security properties (write access to it implies root access to the DB) and it's more likely to already be a k8s secrets mount or equivalent. 
___
*[pkg/server/api_v2_auth.go, line 226 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWesl6W95iVKlGT7jR9:-MWesl6X46lhvIhGkz-v:bs6crav) ([raw file](https://github.com/cockroachdb/cockroach/blob/9079604e145bb0c83d4203ce50f30239b627a534/pkg/server/api_v2_auth.go#L226)):*
> ```Go
> 		err := errors.New("invalid session header")
> 		http.Error(w, err.Error(), http.StatusUnauthorized)
> 		return username, false, nil, err
> ```

You've changed this throughout to return a non-empty username for (some) non-nil errors. All of these methods should have comments indicating when a username may be returned on error and what it means. 
___
*[pkg/server/authentication.go, line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWzZcof5cvNW7Am5CkL:-MWzZcof5cvNW7Am5CkM:b-rcsjv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fd72fa88fb161165002333d98e74c5502ceac19/pkg/server/authentication.go#L150)):*
> ```Go
> 	username, _ := security.MakeSQLUsernameFromUserInput(req.Username, security.UsernameValidation)
> 
> 	// First, try to interpret the password as a predefined cookie. This
> ```

This makes me nervous. I can't point to any specific problems with this approach, but cookies and form parameters have different properties in the web security model and are treated differently in various ways. 

I think I'd prefer to have this as a separate endpoint for emergency logins instead of making the password field perform double duty. It might also be a good idea to limit this functionality to sessions defined in the emergency file instead of all regular web sessions (I don't find the "log in as someone else" scenario in the commit message persuasive). 

(alternately, should the thing in the emergency file be a hashed password instead of a session? that would keep the user experience closer to "standard", although it probably means more invasive changes elsewhere in the login process)
___
*[pkg/server/file_based_authentication.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWf5Xj41B0R_4w0_eom:-MWf5Xj57ZKW5vUi0VtH:ba0j4hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L101)):*
> ```Go
> 		log.Ops.Infof(ctx, "pre-authenticated emergency session found: %v", cookie.ID)
> 	}
> 	if !bytes.Equal(sess.secret, cookie.Secret) {
> ```

I think we probably want to use `crypto/subtle.ConstantTimeCompare` (here and on the standard cookie-validation path if we're not already). And might as well put this after the expiration check. 
___
*[pkg/server/file_based_authentication.go, line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWf6pbMEAVoQvDyPi2Z:-MWf6pbMEAVoQvDyPi2_:bjuej2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L129)):*
> ```Go
> 			select {
> 			case sig := <-ch:
> 				log.Ops.Infof(ctx, "received signal %q, triggering auth file reload", sig)
> ```

We how have three separate log lines emitted on SIGHUP. Maybe this should just be logged once in the signal handler itself (and then the refresher tasks would do their own logging only if they discover changes). 
___
*[pkg/server/file_based_authentication.go, line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWf7L9d41xueJVVRssz:-MWf7L9d41xueJVVRst-:bnu1bm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L151)):*
> ```Go
> 		s.specialSessions.Lock()
> 		defer s.specialSessions.Unlock()
> 		s.specialSessions.sessions = newSessions
> ```

If `hasErrors` is true, we probably want to replace `newSessions` with an empty map instead of using whatever partial data we managed to read. 
___
*[pkg/server/file_based_authentication.go, line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWf7d5H39YVevGjFkE0:-MWf7d5I1RSDC5b1Ogaf:b1bsvhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L157)):*
> ```Go
> 	filename := s.server.cfg.EmergencyAuthenticationSessions
> 	if filename == "" {
> 		log.Ops.Infof(ctx, "emergency access file disabled")
> ```

This will be logged on every SIGHUP? Seems noisy for a feature that the admin explicitly disabled. 
___
*[pkg/server/file_based_authentication.go, line 234 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjAMR9C2Mv346KZMl8:-MWzW-532IJJ-E_F7Mre:b-j6nrlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L234)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes it would with the current implementation. Is that an issue? I am not envisioning this function to be called on a periodic basis. Added a comment to explain.
</blockquote></details>

I think the fact that the same SIGHUP trigger is used for both certificate rotations and reloading this file makes it somewhat more important to control the log noise levels here. It's not unreasonable to reload certs fairly frequently. 
___
*[pkg/server/file_based_authentication.go, line 250 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjDgZrAeADbpIOMep_:-MWzRZxf-NefXcxvvzvQ:bu40x1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L250)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

There is no such reason, but can you explain what the added value would be? The main entropy for this secret is carried by the uuid.NewV4; session Ids are not meant to be very secret, we just expect them to not conflict.

(Which reminds me, we have some risk now here that the session ID generated here will conflict with one in `system.web_sessions`. The UX result of this is that a user who legitimately logs in using a password an a web_sessions cookie can find themselves unable to use that cookie because it doesn't match a per-registered emergency cookie with the same session ID. Do we want to care about this?)
</blockquote></details>

The default should be to use crypto/rand for anything at all security-sensitive; only bring in math/rand when speed is required.

I'm not concerned about the risk of collision in a 64-bit random ID in this context (given that the outcome is a rejection of the non-emergency user rather than granting that user admin powers). 
___
*[pkg/server/file_based_authentication.go, line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWzeS2yDVO1a6jD4YWT:-MWzeS2yDVO1a6jD4YWU:b-jrgobo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fd72fa88fb161165002333d98e74c5502ceac19/pkg/server/file_based_authentication.go#L78)):*
> ```Go
> // maxEmergencyTokenDuration is the maximum validity duration for an
> // emergency token.
> const maxEmergencyTokenDuration = 7 * 24 * time.Hour
> ```

Until we have always-available users (#44134), should we allow "emergency" sessions with arbitrarily long lifetimes? Some users may want to keep the emergency file in place at all times instead of having to run a rarely-used deployment process in the moment. Would there be downsides to doing this? 
___
*[pkg/server/file_based_authentication.go, line 295 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWzbz894J9zdC3i7fPP:-MWzbz894J9zdC3i7fPQ:bvvv1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fd72fa88fb161165002333d98e74c5502ceac19/pkg/server/file_based_authentication.go#L295)):*
> ```Go
> 
> 	authLine := fmt.Sprintf("%s %t %d %s %d",
> 		username, isAdmin, id, uuid, expiryTime.UnixNano())
> ```

The `web_sessions` table contains a hash of the secret, not the secret itself. This is important because it means that stealing a copy of the `web_sessions` table doesn't directly give you the ability to log in. We should do the same for session secrets stored in the emergency file 
___
*[pkg/server/file_based_authentication_test.go, line 182 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWzTBF36TVY_iF62Pfd:-MWzTBF36TVY_iF62Pfe:bnia2su) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication_test.go#L182)):*
> ```Go
> 	}
> 
> 	t.Logf("attempting a new connection with the special credential")
> ```

We need to test some failure cases: Make sure that when the emergency credentials file exist, cookies other than the one contained in the file are rejected. 
___
*[pkg/server/file_based_authentication_test.go, line 173 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWzg3eHAjD24dokatE3:-MWzg3eI7zLShS3LXGeV:b-vjx2bb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0fd72fa88fb161165002333d98e74c5502ceac19/pkg/server/file_based_authentication_test.go#L173)):*
> ```Go
> 
> 	// Write the credential to the file server-side.
> 	if err := ioutil.WriteFile(authFile, []byte(authLine+"\n"), 0644); err != nil {
> ```

Does it fail if that newline's not there? That may be error prone. (do we have any tests of multi-line emergency files?)

We're strict about the permissions of TLS key files; we should probably do the same checks for this file. 


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60723)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @itsbilal, @knz, and @stevendanna)

___
*[pkg/server/authentication.go, line 351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj4gDk1YBjrWvzWPsZ:-MX7Bruf6vxl4MsISza4:b1v8nct) ([raw file](https://github.com/cockroachdb/cockroach/blob/9079604e145bb0c83d4203ce50f30239b627a534/pkg/server/authentication.go#L351)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't know the answer to this question. We have precedent for using multiple return values in other places in this code already, but I am not opposed to adding grouping structs where it makes sense. The error has to remain separate however, and the content of the struct would be undefined if isValid is false. So we'd use a struct just to group `user` and `isAdmin`?

Let me know what you think is valuable.


</blockquote></details>

Given the small number of call sites, probably best to leave it and let the next modification of this code drive the shape of any signature change.
___
*[pkg/server/file_based_authentication.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWf5Xj41B0R_4w0_eom:-MX7AGJE2RkPCUDW69hB:b763zbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L101)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think we probably want to use `crypto/subtle.ConstantTimeCompare` (here and on the standard cookie-validation path if we're not already). And might as well put this after the expiration check. 
</blockquote></details>

Oh nice catch.
___
*[pkg/server/file_based_authentication.go, line 250 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWjDgZrAeADbpIOMep_:-MX7APOWBpHOI-KMNsiO:b-db4dpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0756ff047189fa9a8d75dd6a4900fd3f403681e1/pkg/server/file_based_authentication.go#L250)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The default should be to use crypto/rand for anything at all security-sensitive; only bring in math/rand when speed is required.

I'm not concerned about the risk of collision in a 64-bit random ID in this context (given that the outcome is a rejection of the non-emergency user rather than granting that user admin powers). 
</blockquote></details>

I agree the chance for a conflict is low. But, I still lean towards using crypto/rand just so it is one less thing to think about when reading/auditing this code in the future.
___
*[pkg/server/status.go, line 2412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60723#-MWj55N2-kdBEyRd22uJ:-MX7AZp_1lrasL5qt-sV:b-1b7thx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9079604e145bb0c83d4203ce50f30239b627a534/pkg/server/status.go#L2412)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

As the livestream today demonstrated, we have numerous unit tests that fail when we _mistakenly_ return true here (i.e. we have tests that assert that non-admin sessions don't have access to admin-only stuff).

(The return true happens to be OK here,  but only as long as this function is not called from the wrong place...)
</blockquote></details>

Glad we have that test coverage!


<!-- Sent from Reviewable.io -->

andreimatei commented:
How is `auth-session emergency-login` different from `auth-session login`? Asking because I worry that the need to send a SIGHUP to the nodes will prove impractical in many debug sessions. But I'm thinking that make `auth-session` is enough with the 3rd commit?

knz commented:
- `auth-session login` is the preferred way: it works remotely, but uses `system.web_sessions`  and thus needs a working cluster
- `auth-session emergency-login` is for KV or SQL is unavailable, but does not work fully remotely (needs to change a file + send a signal)

andreimatei commented:
> auth-session emergency-login is for KV or SQL is unavailable, but does not work fully remotely (needs to change a file + send a signal)

I worry about the "change a file and send a signal part". Particularly if it's a file *per node* when you want to access the UI through a balancer? When you're on the phone with someone debugging some crap, telling them to do this I think is often impractical. Not to mention all us us needing to know all these steps.
Could the cli just orchestrate the whole thing though RPCs to all the nodes (or RPC to one node who then fans out to the others)?

andreimatei commented:
I see. It makes sense, but if the always-on user thing is coming, I'd personally question if we need the 2nd commit here. Seems like a lot of code for something that I wonder if it will ever get used. Just my opinion.

knz commented:
> I'd personally question if we need the 2nd commit here.

Well you tell me? (And @tbg too)

I have been working under the assumption since December that "a solution to this is urgent, we're currently blocked from properly investigating stuck clusters since v20.1" from both Tobias, you, Ben and our tech support coworkers.

The always-on users proposal requires more work, in particular refactors within CockroachDB on how the SQL layer fetches users and role membership from system tables. This refactoring (a prerequisite) has been identified as a top priority for the 21.2 roadmap, but it's an internal API change and thus cannot be easily backported.

So I suppose the remaining questions are:

- how important/urgent is it to have a solution?
- do we want a solution here that would work for 20.x and 21.1 cluster via a backport?

Depending on your answer to these questions, I could decided to close this PR and re-prioritize the RFC on the 21.2 roadmap?

knz commented:
(for context, the 2nd commit is precisely the thing that allows you to log in to a cluster that has unavailable ranges)

andreimatei commented:
My opinion is that the fact that we're frequently having trouble logging into "unavailable" clusters is as important and urgent as it gets. That's why I'm very happy you're doing work in the area. At the same time, I personally don't really see this patch as a solution. In the situations where I've run into the difficulty of logging in, I don't think it would have helped - because it seems like too much work during situations where tensions are high, I'm not controlling the keyboard, and everybody is in a hurry. That's just my evaluation of the useful-not useful spectrum; perhaps others will disagree. So, depending on how much code this really is, how invasive, and to what extent this will make it less likely that something better comes along, I personally don't particularly feel the need for this patch (to the extent that I understand it).

Btw, to the extent that I understand commit 3, I like that one for what it does.

andreimatei commented:
> Could the cli just orchestrate the whole thing though RPCs to all the nodes (or RPC to one node who then fans out to the others)?

I'd still come back to this question and ask if the UX of this patch could be made to work a lot better by just some extra work perhaps not that much extra work.

bdarnell commented:
> from both Tobias, you, Ben and our tech support coworkers.

I don't think I've been pushing for this - it's been coming more from folks who are in the trenches debugging things. I agree that we need some solution but I lean towards going straight to the long-term fix with always-available users. I believe there's something we can do there that would be backportable.

I agree with @andreimatei that this is going to be difficult to use during an emergency situation. That's why I suggested removing the duration limitation and making it into a kind of precursor to the always-on-users feature that could be set up as a routine step in advance of an emergency. (although if we're going to be recommending these kinds of routine setup steps we should also be pushing harder for prometheus since external monitoring is always going to be the best solution for debugging an unavailable cluster). 

> Could the cli just orchestrate the whole thing though RPCs to all the nodes (or RPC to one node who then fans out to the others)?

How would the CLI authenticate for this? If the cluster is unavailable, we only have one option available: a client certificate for the `root` user. That would work in some cases, but would be impractical in environments that don't use certificate auth or avoid using the root user. (Or I suppose you could run it from a node and use the node certificate, but I don't think that getting access to a production node's certificate is going to be that much easier than writing a file on that node)

andreimatei commented:
> > Could the cli just orchestrate the whole thing though RPCs to all the nodes (or RPC to one node who then fans out to the others)?

> How would the CLI authenticate for this? If the cluster is unavailable, we only have one option available: a client certificate for the root user. That would work in some cases, but would be impractical in environments that don't use certificate auth or avoid using the root user. (Or I suppose you could run it from a node and use the node certificate, but I don't think that getting access to a production node's certificate is going to be that much easier than writing a file on that node)

Well so how does the patch currently do authentication? I assumed it already needs root certs.

> Or I suppose you could run it from a node and use the node certificate, but I don't think that getting access to a production node's certificate is going to be that much easier than writing a file on that node

I think it would, if you only need to do it on a single node (not all of them), and there's no more signals to be sent and files to be found (other than some certificates). Like, one way or another, I think what we should be shooting for here is that if you have access to the right cert, it's easy to log into the UI. And, separately, I think users should also be able to configure a password on root that can be verified locally by any node (with updates disseminated through gossip or such). I haven't read the always-on users RFC, so y'all might know of much better plans; all I know is that it's quite important for logging in to be EZ.

knz commented:
> Well so how does the patch currently do authentication? I assumed it already needs root certs.

No it does not. It does authentication by requiring write access to the certs directory server-side.

andreimatei commented:
FWIW, in several cases I've observed, opening SQL connections to some node was still working when logging into the console wasn't. So I guess the first thing I'd do is make it such that if I do have the right credentials to open a SQL connections, I can parlay that into a login UI. I think a SQL login still requires a read to `system.users`; it's unclear to me what kinda of authentication does not require a read to `system.users` - but if there's any (for example how does debug.zip authenticate?), I would also like to parlay that into a console login.

bdarnell commented:
The reason web logins are more fragile than sql logins is that we write to the `web_sessions` table, so we require KV write availability (SQL logins are read-only). I think it's worth exploring models in which sessions don't get written to KV (perhaps to gossip instead) even if that comes at the expense of the ability to revoke sessions in certain situations.

knz commented:
So andrei a few things that you'd need to know / understand before you can get into technicalities:

- web authentication uses system.users only the first time around, all requests after that use system.web_sessions
- we can't use cert authentication currently for the HTTP requests easily **because we use different CA and cert chains for SQL and HTTP sessions** (that's a strong ask from customers)
  - we could do this by issuing client certs using the HTTP CA, but we've never done that before
  - would also need to enable TLS client authentication in the HTTP server, which is not enabled today
- most requests need both authentication (system.users / system.web_sessions) and *authorization* ("is this user an admin or not?") which comes from system.role_membership and sometimes system.role_options
- debug zip "seems to work" when it uses the root client cert because we have an authn/authz bypass *for the root cert specifically* and debug zip uses only the RPC port, not HTTP (so the CA/cert chain works)
  - therefore, debug zip's ability to log in is irrelevant to accessing the web UI

knz commented:
> it's worth exploring models in which sessions don't get written to KV (perhaps to gossip instead) even if that comes at the expense of the ability to revoke sessions in certain situations.

yep that's more or less what the always-on users RFC is about

andreimatei commented:
I don't need to engage more in the technicalities beyond this message; I've said my piece.

> therefore, debug zip's ability to log in is irrelevant to accessing the web UI

I find this questionable. You speak of a "strong asks for customers", but I can't help but wonder. At least in an emergency, it seems to me that someone can run a command that uses those root certs (or also the node certs) and gives me a magic string or cookie or something that I can put into the UI login box and have it let me in (like a cluster secret accessible with the root/node certs). Perhaps this is the always-on RFC, in which case I personally would continue to wait for that.

knz commented:
> someone can run a command that uses those root certs (or also the node certs) and gives me a magic string or cookie or something that I can put into the UI login box 

Yep that's a good idea and in fact one I was considering myself early on.  It's not even very complicated to extend this PR in this direction.

The issue I had with that idea is that we have a few customers who restrict access to the RPC port from tech support / operators (using firewall rules), so I felt this wasn't practical. 

However, I also realize we can have both mechanisms available side-by-side: either a file on the filesystem, or a suitably authenticated RPC call from a debug command (the two options don't conflict with each other).

@bdarnell how do you feel about that?

bdarnell commented:
> > it's worth exploring models in which sessions don't get written to KV (perhaps to gossip instead) even if that comes at the expense of the ability to revoke sessions in certain situations.

> yep that's more or less what the always-on users RFC is about

Yes, but I was suggesting something more limited: if Andrei's comment about SQL login being available applies to a lot of the cases we care about, we could implement gossip-based sessions without changing anything about the users table. (if this gives us something more suitable for a backport)

> However, I also realize we can have both mechanisms available side-by-side: either a file on the filesystem, or a suitably authenticated RPC call from a debug command (the two options don't conflict with each other).

I don't like creating multiple backdoor interfaces; it just increases the chance that something will go wrong and we'll leave something wide open (and it doesn't guarantee that we'll cover all customers - why wouldn't someone who restricts network access also restrict filesystem access?). That's especially true for something that's just a stopgap until we can implement a more robust front door option.
